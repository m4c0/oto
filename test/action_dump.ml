let rec print_command = function
  | [] -> print_newline ()
  | [ x ] -> print_endline x
  | x :: xs ->
      print_string x;
      print_string " ";
      print_command xs

let str_from_side (s : Oto.Vm.side) =
  match s with Left -> "left" | Right -> "right" | Middle -> "middle"

let do_action (a : Oto.Vm.action) =
  match a with
  | Background b -> print_command [ "background"; b ]
  | Cast b -> print_command ("cast" :: b)
  | Choose _ -> ()
  | Music b -> print_command [ "music"; b ]
  | Pause -> print_endline "end"
  | Present -> print_endline "present"
  | Speak { side; actor; pose; text } ->
      print_command
        [
          "speak";
          str_from_side side;
          actor;
          Option.value ~default:"none" pose;
          text;
        ]

let rec run (vm : Oto.Vm.t) =
  match vm () with
  | Nil -> print_endline "----"
  | Cons (Choose b, _) ->
      let cb = List.map (fun (c, _) -> c) b in
      print_command ("choose" :: cb);
      List.tl b |> List.hd |> fun (_, nvm) -> run nvm
  | Cons (a, nvm) ->
      do_action a;
      run nvm

let () = run @@ Oto.Vm.from_scene Poc.game
