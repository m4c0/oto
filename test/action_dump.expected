background title
music romance-1
present
background restaurant
music romance-1
speak left lefty none Hello! How are you doing?
speak left lefty none Do you want to know what's happening?
speak right righty none Yeah, you! Baby, baby!
speak middle midly none Click one option to select who you want to date
speak middle midly hot I'm as hot as an idol
choose Lefty Midly Righty
----
pause
background end
music game-over
----
